File: cloudprober/go/scripts/probe.go
// OpenStack API health probe for cloudprober external probe (ONCE mode).
//
// Usage: probe <service>
//   service: identity | compute | network | image | volume
//
// Exit 0  → probe success (HTTP 2xx)
// Exit 1  → probe failure
//
// stdout lines are consumed by cloudprober as additional metrics:
//   http_code <value>
//   auth_failed 1       (only on auth error)
//   not_configured 1    (only when the service URL env var is unset)
//
// Build: go build -o probe probe.go

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	tokenCache = "/tmp/os_token"
	tokenTime  = "/tmp/os_token_time"
)

func maxTokenAge() time.Duration {
	if v := os.Getenv("OS_TOKEN_MAX_AGE"); v != "" {
		if secs, err := strconv.Atoi(v); err == nil {
			return time.Duration(secs) * time.Second
		}
	}
	return 50 * time.Minute
}

func logf(service, format string, args ...any) {
	ts := time.Now().UTC().Format("2006-01-02T15:04:05Z")
	log.Printf("[%s] probe.go [%s] %s", ts, service, fmt.Sprintf(format, args...))
}

// ── Token management ──────────────────────────────────────────────────────────

func getCachedToken() string {
	tb, err := os.ReadFile(tokenTime)
	if err != nil {
		return ""
	}
	cachedAt, err := strconv.ParseInt(strings.TrimSpace(string(tb)), 10, 64)
	if err != nil {
		return ""
	}
	if time.Since(time.Unix(cachedAt, 0)) >= maxTokenAge() {
		return ""
	}
	tok, err := os.ReadFile(tokenCache)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(tok))
}

func refreshToken() (string, error) {
	type domain struct {
		Name string `json:"name"`
	}
	type user struct {
		Name     string `json:"name"`
		Password string `json:"password"`
		Domain   domain `json:"domain"`
	}
	type project struct {
		Name   string `json:"name"`
		Domain domain `json:"domain"`
	}
	type payload struct {
		Auth struct {
			Identity struct {
				Methods  []string `json:"methods"`
				Password struct {
					User user `json:"user"`
				} `json:"password"`
			} `json:"identity"`
			Scope struct {
				Project project `json:"project"`
			} `json:"scope"`
		} `json:"auth"`
	}

	var p payload
	p.Auth.Identity.Methods = []string{"password"}
	p.Auth.Identity.Password.User = user{
		Name:     os.Getenv("OS_USERNAME"),
		Password: os.Getenv("OS_PASSWORD"),
		Domain:   domain{Name: envOrDefault("OS_USER_DOMAIN_NAME", "Default")},
	}
	p.Auth.Scope.Project = project{
		Name:   os.Getenv("OS_PROJECT_NAME"),
		Domain: domain{Name: envOrDefault("OS_PROJECT_DOMAIN_NAME", "Default")},
	}

	body, _ := json.Marshal(p)
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post(
		os.Getenv("OS_AUTH_URL")+"/auth/tokens",
		"application/json",
		bytes.NewReader(body),
	)
	if err != nil {
		return "", fmt.Errorf("keystone request: %w", err)
	}
	defer func() { io.Copy(io.Discard, resp.Body); resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("keystone auth failed: HTTP %d", resp.StatusCode)
	}

	token := resp.Header.Get("X-Subject-Token")
	if token == "" {
		return "", fmt.Errorf("X-Subject-Token header not found")
	}

	if err := os.WriteFile(tokenCache, []byte(token), 0600); err != nil {
		return "", err
	}
	ts := strconv.FormatInt(time.Now().Unix(), 10)
	if err := os.WriteFile(tokenTime, []byte(ts), 0600); err != nil {
		return "", err
	}
	return token, nil
}

func getToken() (string, error) {
	if tok := getCachedToken(); tok != "" {
		return tok, nil
	}
	return refreshToken()
}

// ── Endpoint check ────────────────────────────────────────────────────────────

func checkEndpoint(url, token string) (int, error) {
	timeout := 10
	if v := os.Getenv("OS_PROBE_TIMEOUT"); v != "" {
		if t, err := strconv.Atoi(v); err == nil {
			timeout = t
		}
	}

	client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return 0, err
	}
	if token != "" {
		req.Header.Set("X-Auth-Token", token)
	}

	resp, err := client.Do(req)
	if err != nil {
		return 0, fmt.Errorf("request failed: %w", err)
	}
	defer func() { io.Copy(io.Discard, resp.Body); resp.Body.Close() }()
	return resp.StatusCode, nil
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func envOrDefault(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// ── Main ──────────────────────────────────────────────────────────────────────

type svcConf struct {
	envVar    string
	needsAuth bool
}

var serviceMap = map[string]svcConf{
	"identity": {"OS_AUTH_URL",    false},
	"compute":  {"OS_NOVA_URL",    true},
	"network":  {"OS_NEUTRON_URL", true},
	"image":    {"OS_GLANCE_URL",  true},
	"volume":   {"OS_CINDER_URL",  true},
}

func main() {
	log.SetFlags(0)

	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: probe <identity|compute|network|image|volume>")
		os.Exit(1)
	}

	service := os.Args[1]
	conf, ok := serviceMap[service]
	if !ok {
		fmt.Fprintf(os.Stderr, "Unknown service: %s\n", service)
		os.Exit(1)
	}

	url := os.Getenv(conf.envVar)
	if conf.needsAuth && url == "" {
		logf(service, "URL not configured — skipping")
		fmt.Println("not_configured 1")
		os.Exit(0)
	}

	var token string
	if conf.needsAuth {
		var err error
		token, err = getToken()
		if err != nil {
			logf(service, "authentication error: %v", err)
			fmt.Println("auth_failed 1")
			os.Exit(1)
		}
	}

	httpCode, err := checkEndpoint(url, token)
	if err != nil {
		logf(service, "endpoint error: %v", err)
		fmt.Println("http_code 0")
		os.Exit(1)
	}

	fmt.Printf("http_code %d\n", httpCode)

	if httpCode >= 200 && httpCode < 300 {
		os.Exit(0)
	}

	logf(service, "endpoint returned HTTP %d", httpCode)
	os.Exit(1)
}

File: cloudprober/python/scripts/probe.py
#!/usr/bin/env python3
"""
cloudprober 外部プローブ（ONCEモード）用 OpenStack API 正常性確認スクリプト。

使い方: probe.py <service>
  service: identity | compute | network | image | volume

終了コード:
  0 → プローブ成功（HTTP 2xx）
  1 → プローブ失敗

stdout に出力した行は cloudprober の追加メトリクスとして取り込まれる:
  http_code <値>
  auth_failed 1       （認証エラー時のみ）
  not_configured 1    （サービスURLの環境変数が未設定の場合のみ）
"""

import os
import sys
import time

import requests

# トークンのファイルキャッシュパス（サブプロセス再起動をまたいで再利用する）
TOKEN_CACHE = "/tmp/os_token"
TOKEN_TIME  = "/tmp/os_token_time"

# キャッシュ有効期限（秒）。OpenStack トークンのデフォルト寿命は1時間なので50分を既定値とする
MAX_TOKEN_AGE = int(os.environ.get("OS_TOKEN_MAX_AGE", "3000"))


def log(service: str, msg: str) -> None:
    """標準エラー出力にタイムスタンプ付きのログを書き出す。"""
    ts = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    print(f"[{ts}] probe.py [{service}] {msg}", file=sys.stderr)


# ── トークン管理 ──────────────────────────────────────────────────────────────

def get_cached_token() -> str | None:
    """キャッシュファイルが存在し有効期限内であればトークンを返す。"""
    try:
        with open(TOKEN_TIME) as f:
            age = int(time.time()) - int(f.read().strip())
        if age < MAX_TOKEN_AGE:
            with open(TOKEN_CACHE) as f:
                return f.read().strip()
    except (FileNotFoundError, ValueError):
        # キャッシュファイルが存在しない、または破損している場合は再取得する
        pass
    return None


def refresh_token() -> str:
    """Keystone v3 パスワード認証でトークンを取得し、キャッシュファイルに保存する。"""
    # Keystone v3 認証リクエストのペイロードを組み立てる
    payload = {
        "auth": {
            "identity": {
                "methods": ["password"],
                "password": {
                    "user": {
                        "name":     os.environ["OS_USERNAME"],
                        "password": os.environ["OS_PASSWORD"],
                        # ユーザードメインが未指定の場合は "Default" を使用する
                        "domain": {"name": os.environ.get("OS_USER_DOMAIN_NAME", "Default")},
                    }
                },
            },
            "scope": {
                "project": {
                    "name": os.environ["OS_PROJECT_NAME"],
                    # プロジェクトドメインが未指定の場合は "Default" を使用する
                    "domain": {"name": os.environ.get("OS_PROJECT_DOMAIN_NAME", "Default")},
                }
            },
        }
    }

    resp = requests.post(
        os.environ["OS_AUTH_URL"] + "/auth/tokens",
        json=payload,
        timeout=30,
    )
    # HTTP 4xx / 5xx の場合は例外を送出して呼び出し元でキャッチする
    resp.raise_for_status()

    # トークンはレスポンスボディではなくヘッダー X-Subject-Token に含まれる
    token = resp.headers["X-Subject-Token"]

    # 次回のプローブ実行で再利用できるようにファイルへ保存する
    with open(TOKEN_CACHE, "w") as f:
        f.write(token)
    with open(TOKEN_TIME, "w") as f:
        f.write(str(int(time.time())))

    return token


def get_token() -> str:
    """キャッシュが有効であればそれを返し、期限切れの場合は再取得する。"""
    token = get_cached_token()
    return token if token else refresh_token()


# ── エンドポイント確認 ────────────────────────────────────────────────────────

def check_endpoint(url: str, token: str | None) -> int:
    """指定 URL に GET リクエストを送り HTTP ステータスコードを返す。"""
    timeout = int(os.environ.get("OS_PROBE_TIMEOUT", "10"))
    # identity サービスは認証不要のためトークンがない場合はヘッダーを付与しない
    headers = {"X-Auth-Token": token} if token else {}
    resp = requests.get(url, headers=headers, timeout=timeout)
    return resp.status_code


# ── サービス定義 ──────────────────────────────────────────────────────────────

# キー: サービス名, 値: (URLを格納する環境変数名, 認証トークンが必要か)
SERVICE_MAP = {
    "identity": ("OS_AUTH_URL",    False),  # Keystone はパブリックエンドポイントのため認証不要
    "compute":  ("OS_NOVA_URL",    True),
    "network":  ("OS_NEUTRON_URL", True),
    "image":    ("OS_GLANCE_URL",  True),
    "volume":   ("OS_CINDER_URL",  True),
}


# ── エントリポイント ──────────────────────────────────────────────────────────

def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: probe.py <identity|compute|network|image|volume>", file=sys.stderr)
        return 1

    service = sys.argv[1]
    if service not in SERVICE_MAP:
        print(f"Unknown service: {service}", file=sys.stderr)
        return 1

    env_var, needs_auth = SERVICE_MAP[service]
    url = os.environ.get(env_var, "")

    # URLが未設定のサービスはスキップ扱いとし、アラートを発生させない
    if needs_auth and not url:
        log(service, "URL not configured — skipping")
        print("not_configured 1")
        return 0

    # 認証が必要なサービスは Keystone からトークンを取得する
    token = None
    if needs_auth:
        try:
            token = get_token()
        except Exception as e:
            log(service, f"authentication error: {e}")
            print("auth_failed 1")
            return 1

    # エンドポイントへの疎通確認を実行する
    try:
        http_code = check_endpoint(url, token)
    except Exception as e:
        log(service, f"endpoint error: {e}")
        print("http_code 0")
        return 1

    # HTTP ステータスコードを cloudprober のメトリクスとして出力する
    print(f"http_code {http_code}")

    # 2xx であれば正常、それ以外は異常と判定する
    if 200 <= http_code < 300:
        return 0

    log(service, f"endpoint returned HTTP {http_code}")
    return 1


if __name__ == "__main__":
    sys.exit(main())
